File: resilient/resilient-annotation-processor/src/main/java/io/koraframework/resilient/annotation/processor/aop/CircuitBreakerKoraAspect.java
package io.koraframework.resilient.annotation.processor.aop;

import com.palantir.javapoet.*;
import io.koraframework.annotation.processor.common.CommonClassNames;
import io.koraframework.annotation.processor.common.MethodUtils;
import io.koraframework.annotation.processor.common.ProcessingErrorException;
import io.koraframework.aop.annotation.processor.KoraAspect;

import javax.annotation.processing.ProcessingEnvironment;
import javax.lang.model.element.AnnotationMirror;
import javax.lang.model.element.ExecutableElement;
import javax.lang.model.element.TypeElement;
import javax.lang.model.type.TypeMirror;
import java.util.Optional;
import java.util.Set;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionException;

import static com.palantir.javapoet.CodeBlock.joining;

public class CircuitBreakerKoraAspect implements KoraAspect {

    private static final ClassName ANNOTATION_TYPE = ClassName.get("io.koraframework.resilient.circuitbreaker.annotation", "CircuitBreakable");
    private static final ClassName CIRCUIT_BREAKER = ClassName.get("io.koraframework.resilient.circuitbreaker", "CircuitBreaker");
    private static final ClassName PERMITTED_EXCEPTION = ClassName.get("io.koraframework.resilient.circuitbreaker.exception", "CallNotPermittedException");

    private final ProcessingEnvironment env;

    public CircuitBreakerKoraAspect(ProcessingEnvironment env) {
        this.env = env;
    }

    @Override
    public Set<ClassName> getSupportedAnnotationClassNames() {
        return Set.of(ANNOTATION_TYPE);
    }

    @Override
    public ApplyResult apply(ExecutableElement method, String superCall, AspectContext aspectContext) {
        if (MethodUtils.isPublisher(method)) {
            throw new ProcessingErrorException("@%s can't be applied for type ".formatted(ANNOTATION_TYPE.simpleName()) + CommonClassNames.publisher, method);
        } else if(MethodUtils.isFuture(method)) {
            throw new ProcessingErrorException("@%s can't be applied for type ".formatted(ANNOTATION_TYPE) + method.getReturnType().toString(), method);
        }

        final Optional<? extends AnnotationMirror> mirror = method.getAnnotationMirrors().stream()
            .filter(a -> a.getAnnotationType().toString().equals(ANNOTATION_TYPE.canonicalName()))
            .findFirst();
        final TypeMirror circuitBreakerTypeMirror = mirror.flatMap(a -> a.getElementValues().entrySet().stream()
                .filter(e -> e.getKey().getSimpleName().contentEquals("value"))
                .map(e -> (TypeMirror) e.getValue().getValue())
                .findFirst())
            .orElseThrow();

        var circuitBreakerElement = (TypeElement) env.getTypeUtils().asElement(circuitBreakerTypeMirror);
        var baseCircuitBreakerType = env.getElementUtils().getTypeElement(CIRCUIT_BREAKER.canonicalName()).asType();
        if (!env.getTypeUtils().isAssignable(circuitBreakerTypeMirror, baseCircuitBreakerType)) {
            throw new ProcessingErrorException("@%s value must extend %s".formatted(ANNOTATION_TYPE.simpleName(), CIRCUIT_BREAKER.canonicalName()), method);
        }

        var circuitType = env.getTypeUtils().getDeclaredType(circuitBreakerElement);
        final String fieldCircuit = aspectContext.fieldFactory().constructorParam(circuitType, java.util.List.of());

        final CodeBlock body;
        if (MethodUtils.isCompletionStage(method)) {
            body = buildBodyCompletableStage(method, superCall, fieldCircuit);
        } else {
            body = buildBodySync(method, superCall, fieldCircuit);
        }

        return new ApplyResult.MethodBody(body);
    }

    private CodeBlock buildBodySync(ExecutableElement method, String superCall, String cbField) {
        final CodeBlock superMethod = buildMethodCall(method, superCall);
        final CodeBlock methodCall = MethodUtils.isVoid(method)
            ? superMethod
            : CodeBlock.of("var _result = $L", superMethod.toString());

        final CodeBlock returnCall = MethodUtils.isVoid(method)
            ? CodeBlock.of("return")
            : CodeBlock.of("return _result");

        return CodeBlock.builder().add("""
            try {
                $L.acquire();
                $L;
                $L.releaseOnSuccess();
                $L;
            } catch ($T _e) {
                throw _e;
            } catch (Throwable _e) {
                $L.releaseOnError(_e);
                throw _e;
            }
            """, cbField, methodCall.toString(), cbField, returnCall.toString(), PERMITTED_EXCEPTION, cbField).build();
    }

    private CodeBlock buildBodyCompletableStage(ExecutableElement method, String superCall, String cbField) {
        final CodeBlock superMethod = buildMethodCall(method, superCall);

        return CodeBlock.builder().add("""
            try {
                $L.acquire();
                return $L.whenComplete((_r, _e) -> {
                    if (_e != null) {
                        if (_e instanceof $T ce) {
                            _e = ce.getCause();
                        }
                        $L.releaseOnError(_e);
                    } else {
                        $L.releaseOnSuccess();
                    }
                });
            } catch ($T _e) {
                return $T.failedFuture(_e);
            } catch (Throwable _e) {
                $L.releaseOnError(_e);
                throw _e;
            }
            """, cbField, superMethod, CompletionException.class, cbField, cbField, PERMITTED_EXCEPTION, CompletableFuture.class, cbField).build();
    }

    private CodeBlock buildMethodCall(ExecutableElement method, String call) {
        return method.getParameters().stream().map(p -> CodeBlock.of("$L", p)).collect(joining(", ", call + "(", ")"));
    }
}

File: resilient/resilient-symbol-processor/src/main/kotlin/io/koraframework/resilient/symbol/processor/aop/CircuitBreakerKoraAspect.kt
package io.koraframework.resilient.symbol.processor.aop

import com.google.devtools.ksp.getClassDeclarationByName
import com.google.devtools.ksp.processing.Resolver
import com.google.devtools.ksp.symbol.KSFunctionDeclaration
import com.google.devtools.ksp.symbol.KSType
import com.squareup.kotlinpoet.ClassName
import com.squareup.kotlinpoet.CodeBlock
import com.squareup.kotlinpoet.MemberName
import com.squareup.kotlinpoet.ksp.toTypeName
import io.koraframework.aop.symbol.processor.KoraAspect
import io.koraframework.ksp.common.AnnotationUtils.findAnnotation
import io.koraframework.ksp.common.AnnotationUtils.findValue
import io.koraframework.ksp.common.CommonClassNames
import io.koraframework.ksp.common.FunctionUtils.isCompletionStage
import io.koraframework.ksp.common.FunctionUtils.isFlow
import io.koraframework.ksp.common.FunctionUtils.isFlux
import io.koraframework.ksp.common.FunctionUtils.isFuture
import io.koraframework.ksp.common.FunctionUtils.isMono
import io.koraframework.ksp.common.FunctionUtils.isSuspend
import io.koraframework.ksp.common.FunctionUtils.isVoid
import io.koraframework.ksp.common.exception.ProcessingErrorException
import java.util.concurrent.CompletionStage
import java.util.concurrent.Future

class CircuitBreakerKoraAspect(val resolver: Resolver) : KoraAspect {

    companion object {
        private val ANNOTATION_TYPE = ClassName("io.koraframework.resilient.circuitbreaker.annotation", "CircuitBreakable")
        private val CIRCUIT_BREAKER = ClassName("io.koraframework.resilient.circuitbreaker", "CircuitBreaker")
        private val PERMITTED_EXCEPTION = ClassName("io.koraframework.resilient.circuitbreaker.exception", "CallNotPermittedException")
    }

    override fun getSupportedAnnotationTypes(): Set<String> {
        return setOf(ANNOTATION_TYPE.canonicalName)
    }

    override fun apply(ksFunction: KSFunctionDeclaration, superCall: String, aspectContext: KoraAspect.AspectContext): KoraAspect.ApplyResult {
        if (ksFunction.isFuture()) {
            throw ProcessingErrorException("@CircuitBreakable can't be applied for types assignable from ${Future::class.java}", ksFunction)
        } else if (ksFunction.isCompletionStage()) {
            throw ProcessingErrorException("@CircuitBreakable can't be applied for types assignable from ${CompletionStage::class.java}", ksFunction)
        } else if (ksFunction.isMono()) {
            throw ProcessingErrorException("@CircuitBreakable can't be applied for types assignable from ${CommonClassNames.mono}", ksFunction)
        } else if (ksFunction.isFlux()) {
            throw ProcessingErrorException("@CircuitBreakable can't be applied for types assignable from ${CommonClassNames.flux}", ksFunction)
        }

        val circuitBreakerType = ksFunction.findAnnotation(ANNOTATION_TYPE)!!
            .findValue<KSType>("value")!!
        val baseCircuitBreaker = resolver.getClassDeclarationByName(CIRCUIT_BREAKER.canonicalName)!!.asStarProjectedType()
        if (!baseCircuitBreaker.isAssignableFrom(circuitBreakerType)) {
            throw ProcessingErrorException("@CircuitBreakable value must extend ${CIRCUIT_BREAKER.canonicalName}", ksFunction)
        }

        val fieldCircuit = aspectContext.fieldFactory.constructorParam(
            circuitBreakerType.toTypeName(),
            listOf()
        )

        val body = if (ksFunction.isFlow()) {
            buildBodyFlow(ksFunction, superCall, fieldCircuit)
        } else if (ksFunction.isSuspend()) {
            buildBodySuspend(ksFunction, superCall, fieldCircuit)
        } else {
            buildBodySync(ksFunction, superCall, fieldCircuit)
        }
        return KoraAspect.ApplyResult.MethodBody(body)
    }

    private fun buildBodySync(
        method: KSFunctionDeclaration, superCall: String, fieldCircuitBreaker: String
    ): CodeBlock {
        val superMethod = buildMethodCall(method, superCall)
        val methodCall = if(method.isVoid()) superMethod else CodeBlock.of("val t = %L", superMethod)
        val returnCall = if(method.isVoid()) CodeBlock.of("") else CodeBlock.of("t")

        return CodeBlock.builder().add(
            """
            return try {
                %L.acquire()
                %L
                %L.releaseOnSuccess()
                %L
            } catch (e: %T) {
                throw e
            } catch (e: Throwable) {
                %L.releaseOnError(e)
                throw e
            }
            """.trimIndent(), fieldCircuitBreaker, methodCall, fieldCircuitBreaker,
            returnCall, PERMITTED_EXCEPTION, fieldCircuitBreaker
        ).build()
    }

    private fun buildBodyFlow(
        method: KSFunctionDeclaration, superCall: String, fieldCircuitBreaker: String
    ): CodeBlock {
        val flowMember = MemberName("kotlinx.coroutines.flow", "flow")
        val emitMember = MemberName("kotlinx.coroutines.flow", "emitAll")
        val superMethod = buildMethodCall(method, superCall)
        return CodeBlock.builder().add(
            """
            return %M {
                try {
                    %L.acquire()
                    %M(%L)
                    %L.releaseOnSuccess()
                } catch (e: %T) {
                    throw e
                } catch (e: Throwable) {
                    %L.releaseOnError(e)
                    throw e
                }
            }
            """.trimIndent(), flowMember, fieldCircuitBreaker, emitMember, superMethod.toString(),
            fieldCircuitBreaker, PERMITTED_EXCEPTION, fieldCircuitBreaker
        ).build()
    }

    private fun buildBodySuspend(
        method: KSFunctionDeclaration, superCall: String, fieldCircuitBreaker: String
    ): CodeBlock {
        val superMethod = buildMethodCall(method, superCall)
        val methodCall = if (method.isVoid()) superMethod else CodeBlock.of("val t = %L", superMethod)
        val returnCall = if (method.isVoid()) CodeBlock.of("") else CodeBlock.of("t")

        return CodeBlock.builder().add(
            """
            return try {
                %L.acquire()
                %L
                %L.releaseOnSuccess()
                %L
            } catch (e: %T) {
                throw e
            } catch (e: Throwable) {
                %L.releaseOnError(e)
                throw e
            }
            """.trimIndent(), fieldCircuitBreaker, methodCall, fieldCircuitBreaker,
            returnCall, PERMITTED_EXCEPTION, fieldCircuitBreaker
        ).build()
    }

    private fun buildMethodCall(method: KSFunctionDeclaration, call: String): CodeBlock {
        return CodeBlock.of(method.parameters.asSequence().map { p -> CodeBlock.of("%L", p) }.joinToString(", ", "$call(", ")"))
    }
}
